core/block.js

```diff
@@ -668,6 +670,45 @@ Blockly.Block.prototype.setMovable = function(movable) {
   this.movable_ = movable;
 };
 
+/**
+ * Get whether is block is duplicatable or not. If duplicating this block and
+ * descendants will put this block over the workspace's capacity this block is
+ * not duplicatable. If duplicating this block and decendents will put any
+ * type over their maxInstances this block is not duplicatable.
+ * @return {boolean} True if duplicatable.
+ */
+Blockly.Block.prototype.isDuplicatable = function() {
+  if (!this.workspace.hasBlockLimits()) {
+    return true;
+  }
+
+  var copyableBlocks = this.getDescendants(true);
+  // Remove all "next statement" blocks because they will not be copied.
+  var nextBlock = this.getNextBlock();
+  if (nextBlock) {
+    var index = copyableBlocks.indexOf(nextBlock);
+    copyableBlocks.splice(index, copyableBlocks.length - index);
+  }
+
+  if (copyableBlocks.length > this.workspace.remainingCapacity()) {
+    return false;
+  }
+
+  var copyableBlocksTypeCounts = {};
```

Comment:
Object.create(null) would be better than {}, since if there's ever a block type named "constructor", or "toString", the results would be unexpected.